content/dharma/abstract-data-structures.md

```diff
@@ -0,0 +1,118 @@
+---
+Title: "Introduction to ADTs in JavaScript"
+Description: "An introduction to Abstract Data Types"
+DatePublished: "2022-04-25"
+Categories:
+  - "code-foundations"
+  - "computer-science"
+Tags:
+  - "Data Structures"
+  - "Classes"
+  - "OOP"
+CatalogContent:
+  - "paths/computer-science"
+  - "introduction-to-javascript"
+---
+
+![Introduction to Abstract Data Type](./adt.png)
+
+## Introduction
+
+An Abstract Data Type, as the name suggests, is an abstract understanding of a data structure. An **Abstract Data Type** is defined through its behavior and characteristics, particularly in terms of what data can be stored into it, the operations that can be performed on this data, and the behavior of these operations. For example, stacks and queues can be internally implemented using linked-lists made up of nodes or lists/arrays. However, the primary function of a stack is to be a Last In First Out(LIFO) data structure and the primary function of a queue is to be a First In First Out(FIFO) data structure. The behavior, from the point of the user, remains intact, regardless of the internal implementation either using linked-lists or arrays/lists. If the user was interacting with a stack, the user will simply worry about pushing data onto the stack or popping data off the stack. The user won't need to have the knowledge how that stack is working internally.
+
+In contrast to the data structures which are specific and detailed implementations that deal with how the data structure does its job, abstract data type focuses on _what it does_ and not how it does its job. In short, the ADT defines what that particular data construct must do and the data structure is the concrete implementation of that construct.
+
+An analogy to explain abstract data type in terms of web development would be CRUD APIs. The user of any CRUD API has to simply know what request method(GET, POST, etc.) should they send and if they followed the rules of the API, the API server would send data back. The user didn't have to worry about the internal workings of the API server. The user simply had to know the rules of interactions and behavior of CRUD API. In this case, the CRUD API is functioning as an abstract data type _from the perspective of the user_.
```

Comment:
```suggestion
An analogy to explain abstract data types in terms of web development would be CRUD APIs. The user of any CRUD API has to simply know what request method (GET, POST, etc.) should they send and if they followed the rules of the API, the API server would send data back. The user didn't have to worry about the internal workings of the API server. The user simply had to know the rules of interactions and behavior of a CRUD API. In this case, the CRUD API is functioning as an abstract data type _from the perspective of the user_.
```

Reply:
You may also want to spell out CRUD initially.

Reply:
Ah. I'll add that as well.

Reply:
With regards to the first sentence in this line, I would change "abstract data types" to "ADTs". Going back to @SSwiniarski's [earlier](https://github.com/Codecademy/ugc/pull/51/files#r862357897) and [subsequent](https://github.com/Codecademy/ugc/pull/51/files#r862359520) suggestions, this abbreviation should be introduced at the first reference to "Abstract Data Type(s)" and exclusively used afterwards.